Speed-up CI by reorganizing tests

alalazo opened the pull request:
This PR tries to reorganize how we run tests to:
1. Run less tests in PRs that modify only package recipes
2. Introduce dependencies between tests, so that fast tests passing are a prerequisite to run slow tests

to counter the raise in the waiting time seen in the last few days for PRs to start running on Github Actions. 

Modifications:
- [x] Both linux and MacOS unit tests are in the same workflow, and are conditional on style tests passing
- [x] PRs that change only package files skip running slow tests and avoid submitting coverage
- [x] MacOS unit tests fail fast (exit on first error)

Relevant links:
- https://github.community/t/how-do-i-specify-job-dependency-running-in-another-workflow/16482/15
- https://github.community/t/feature-request-conditional-required-checks/16761/21

alalazo committed:
unit tests: mark slow tests as "maybeslow"

This commit also removes the "network" marker and
marks every "network" test as "maybeslow". Tests
marked as db are maintained, but they're not slow
anymore.

alalazo committed:
GA: require style tests to pass before running unit-tests

cosmicexplorer commented on .github/workflows/linux_unit_tests.yaml:
```diff
@@ -10,7 +10,67 @@ on:
+        pip install --upgrade pip
+        pip install --upgrade vermin
+    - name: Minimum Version (Spack's Core)
+      run: vermin --backport argparse --backport typing -t=2.6- -t=3.5- -v lib/spack/spack/ lib/spack/llnl/ bin/
```
This is orthogonal to CI performance, but the vermin output is extremely annoying and time-consuming to map back to actual source of incompatibility. I think it's probably worth creating a wrapper script (I can do this, in a separate PR) which runs vermin with `-vvv` to get the offending lines and columns, and avoids printing files which have 0 compatibility problems. Does that make sense?

The `vermin` cli might have these options instead of needing a wrapper script, but I don't want to block this PR on any `vermin` things.

adamjstewart replied:
vermin cli doesn't have those options, but I wish it did. The vermin developer knows about Spack and contributes new versions when he releases them, so it may be easier to contribute directly to vermin instead of writing a wrapper script.

cosmicexplorer commented on .github/workflows/linux_unit_tests.yaml:
```diff
@@ -106,14 +168,15 @@ jobs:
     - name: Run shell tests
       env:
           COVERAGE: true
+          PYTEST_ADDOPTS: '-m "not maybeslow"'
```
This marker seems stronger than just "maybe" now that you've taken the time to add it to more tests. Do we want to rename it to something like "long-running" (or just "slow")? Also, I think there may be tiers of slowness within these tests -- perhaps:
- < 1 min
- 1-2 min
- 3+ min

What do you think? Or should we focus on only the "slow" vs "not slow" distinction first?

```suggestion
          PYTEST_ADDOPTS: '-m "not longrunning"'
```

cosmicexplorer replied:
Also, I don't know if we're using parallel pytest processes at all? I know I mention pants a lot, but this is one thing pants is *really* good at: automatically running python tests in parallel with per-file caching. See https://www.pantsbuild.org/docs/python-test-goal. We would not need to modify spack's deployment model or change any code whatsoever for this to work -- we just need to add the pants runner script and a `pants.toml` file, and pants will work without modifying any python code. Please let me know if this is something I can investigate.

alalazo replied:
> Or should we focus on only the "slow" vs "not slow" distinction first?

I would say so. Right now I think ~150 tests out of ~2700 are responsible for 50% of the testing time. I'm absolutely fine using "slow" instead of "maybeslow" if people prefer that.

alalazo replied:
> Also, I don't know if we're using parallel pytest processes at all?

No we don't. Some of our tests are not properly isolated (which is unfortunate, I know). Running in parallel is something I hope we'll be able to do, but I think it currently requires some effort to track the tests that can step on each other and fix them. Also, I tried to run tests in parallel in the past only via `pytest-xdist` so my experience refers to that.

cosmicexplorer commented on lib/spack/spack/test/cmd/versions.py:
```diff
@@ -24,50 +23,50 @@ def test_safe_versions():
 
 
-@pytest.mark.network
+@pytest.mark.maybeslow
```
Would it be useful to leave the `.network` mark on as well? It seems like useful metadata.

alalazo committed:
GA: make MacOS unit tests fail fast

alalazo committed:
GA: move all unit tests into the same workflow, run style tests as a prerequisite

All the unit tests have been moved into the same workflow so that a single
run of the dorny/paths-filter action can be used to ask for coverage based
on the files that have been changed in a PR. The basic idea is that for PRs
that introduce only changes to packages coverage is not necessary, this
resulting in a faster execution of the tests.

Also, for package only PRs slow unit tests are skipped.

Finally, MacOS and linux unit tests are now conditional on style tests passing
meaning that e.g. we won't waste a MacOS worker if we know that the PR has
flake8 issues.

tgamblin commented on .github/workflows/unit_tests.yaml:
```diff
@@ -0,0 +1,338 @@
+      run: |
+        pip install --upgrade pip
+        pip install --upgrade vermin
+    - name: Minimum Version (Spack's Core)
```
nitpick: call this `vermin` or `python version` -- it sounds like it's checking minimum version of spack or of the packages but it's really python support.

tgamblin commented on .github/workflows/unit_tests.yaml:
```diff
@@ -0,0 +1,338 @@
+    # setting environment variables from earlier steps: https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable
+    #
+    - id: coverage
+      # Run the following jobs with coverage if this is a PR and core has been modified
```
nitpick: "following" sounds like you mean something in the text immediately following this line -- I think "subsequent" would make it clearer that this is about the next jobs in the workflow (could be just me).

tgamblin commented on share/spack/qa/run-shell-tests:
```diff
@@ -18,7 +18,11 @@
+if [ "$COVERAGE" = true ]; then
+  check_dependencies $coverage kcov git hg svn
+else
+  echo "ONLY PACKAGE RECIPES CHANGED [skipping coverage]"
```
I think we do not really know here why coverage is being skipped -- maybe just output "COVERAGE not set to 'true'. skipping coverage"

tgamblin reviewed (dismissed):
LGTM!   I have some minor comments but feel free to merge this!

alalazo committed:
Addressed review comments

alalazo committed:
Skipping slow tests on MacOS for package only recipes

alalazo committed:
QA: make tests on changes correct before merging

alalazo commented:
@tgamblin Ready to be merged. There's a transient failure on rhel8 (see https://github.com/spack/spack/issues/17405) but I'd merge without retriggering since that run passed multiple times already in previous commits.

haampie commented:
This was very much welcome... :sweat_smile:
